### PR TITLE
Add Terraform config with CloudBuild triggers for building docker images and wheels

### PR DESCRIPTION
**Pull request by mateuszlewko**
This PR adds pieces missing from https://github.com/pytorch/xla/pull/4531.

This Terraform setup provisions:
- public storage bucket for PyTorch and PyTorch/XLA wheels.
- private storage bucket for Terraform state.
- public artifact repository for docker images.
- cloud builds for nightly and release docker images and wheels.
- schedule jobs and a service account for triggering cloud build.

This setup provisions resource in the tpu-pytorch project, but appends suffix "-staging" to many resources to distinguish them. It doesn't reuse / import any existing resources.

The most important and user visible part is configuration of docker images and wheels in `docker/experimental/terraform_cloudbuild/vars/staging.tfvars`:

```terraform
docker_images = [
  {
    image = "development"
    description = "Build development image with TPU support."
    dockerfile = "development.Dockerfile"
    trigger_files = [
      "docker/experimental/ansible/**",
      "docker/experimental/terraform_cloudbuild/**",
    ]
    build_args = {
      python_version="3.8"
      arch="amd64"
      accelerator="tpu"
    }
    image_tags = [
      "tpu_amd64",
      "tpu_amd64_$(date +%Y%m%d)",
    ]
  },
  {
    name = "xla-nightly-38-cuda11-8"
    image = "xla"
    description = "Build nightly image with CUDA support"
    trigger_schedule = "0 0 * * *"
    build_args = {
      python_version="3.8"
      arch="amd64"
      accelerator="cuda"
      cuda_version="11.8"
    }
    image_tags = [
      "nightly_3.8_cuda11.8",
      "nightly_3.8_cuda11.8_$(date +%Y%m%d)",
    ]
    wheels = ["/wheels/*.whl"]
  },
  {
    name = "xla-nightly-38-tpu"
    image = "xla"
    description = "Build nightly image with TPU support"
    trigger_schedule = "0 0 * * *"
    build_args = {
      python_version="3.8"
      arch="amd64"
      accelerator="tpu"
    }
    image_tags = [
      "nightly_3.8_tpuvm",
      "nightly_3.8_$(date +%Y%m%d)",
    ]
    wheels = ["/wheels/*.whl"]
  },
  {
    name = "xla-1-13-38-tpu"
    image = "xla"
    description = "Build v1.13.0 image with TPU support"
    # Don't use this tag here, since the repositiory at version v1.13.0
    # doesn't contain ansible setup.
    # git_tag = "v1.13.0"
    trigger_schedule = "0 0 * * *"
    build_args = {
      python_version="3.8"
      arch="amd64"
      accelerator="tpu"
      # Fetch sources at the desired tag.
      pytorch_git_rev="v1.13.0"
      xla_git_rev="v1.13.0"
    }
    image_tags = [
      "r1.13_3.8_tpuvm",
    ]
    wheels = ["/wheels/*.whl"]
  },
  {
    name = "xla-1-12-38-tpu"
    image = "xla"
    description = "Build v1.12.0 image with TPU support"
    # git_tag = "v1.12.0"
    trigger_schedule = "0 0 * * *"
    build_args = {
      python_version="3.8"
      arch="amd64"
      accelerator="tpu"
      pytorch_git_rev="v1.13.0"
      xla_git_rev="v1.13.0"
    }
    image_tags = [
      "r1.12_3.8_tpuvm",
    ]
    wheels = ["/wheels/*.whl"]
  },
]
```

### CODE REVIEW

File: docker/experimental/ansible/config/vars.yaml [hunk 1]
```diff
@@ -1,7 +1,10 @@
 # Used for fetching cuda from the right repo, see apt.yaml.
 cuda_repo: ubuntu1804
+cuda_version: 11.8
 # Used for fetching clang from the right repo, see apt.yaml.
 llvm_debian_repo: buster
-clang_version: 11
+clang_version: 10
```
Comment (stgpetrovic) [inline docker/experimental/ansible/config/vars.yaml:6]: It seems that we're going for gcc in the end until clang/bazel interaction is fixed. We should keep clang around though for other things so nothing to change here, just fyi.

File: docker/experimental/ansible/roles/fetch_srcs/tasks/main.yaml [hunk 2]
```diff
@@ -34,8 +34,10 @@
     # localhost.
     remote_src: true
     strip: 1
+    ignore_whitespace: true
     basedir: "{{ (src_root, 'pytorch/xla/third_party/tensorflow') | path_join }}"
```
Comment (stgpetrovic) [inline docker/experimental/ansible/roles/fetch_srcs/tasks/main.yaml:38]: This is now gone, and bazel does the patching.

**Review by stgpetrovic — COMMENTED** [review]
Nice!

**Review by will-cromar — COMMENTED** [review]
Thanks!

In general, feel free to overwrite or delete the existing content in `docker/experimental` instead of creating new copies of the Docker build and Terraform triggers. The experimental build is still experimental, so it's okay if it breaks while we transition to something new.

File: docker/experimental/ansible/config/vars.yaml (hunk 1 above)
Comment (mateuszlewko) [reply docker/experimental/ansible/config/vars.yaml:6]: Ack

**Comment by mateuszlewko** [comment]
> Thanks!
> 
> In general, feel free to overwrite or delete the existing content in `docker/experimental` instead of creating new copies of the Docker build and Terraform triggers. The experimental build is still experimental, so it's okay if it breaks while we transition to something new.

Done: moved all my changes to /docker/experimental/terraform

File: docker/experimental/terraform_cloudbuild/vars/staging.tfvars [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    }
+    image_tags = [
+      "nightly_3.8_tpuvm",
+      "nightly_3.8_$(date +%Y%m%d)",
+    ]
+    wheels = ["/wheels/*.whl"]
+    timeout_m = 60 * 6
+  },
+  {
+    trigger_name = "xla-1-13-38-tpu"
```
[the remark this answers is not in the document]
Comment (mateuszlewko) [reply docker/experimental/terraform_cloudbuild/vars/staging.tfvars]: Done

File: docker/experimental/ansible/roles/fetch_srcs/tasks/main.yaml (hunk 2 above)
Comment (mateuszlewko) [reply docker/experimental/ansible/roles/fetch_srcs/tasks/main.yaml:38]: But it's not gone for v1.12 and v.1.13 releases? The problem with this approach is that build & deployment configuration is not bound with sources version (and for now I don't think we can improve on that). So this build process has to work for current and future versions of the code as well as older ones. 

Do you think that having this step while bazel is doing the patching is problematic? I added `ignore_errors: true` so if the step fails it's fine, the playbook still continues.

File: docker/experimental/terraform_cloudbuild/vars/staging.tfvars (hunk 3 above)
[the remark this answers is not in the document]
Comment (mateuszlewko) [reply docker/experimental/terraform_cloudbuild/vars/staging.tfvars]: However, there is no 2.0.0 tag for PyTorch/XLA project, only for PyTorch. What should be git revision for PyTorch/XLA?

**Pull request merged**